### PR TITLE
ORC-310 better error handling for codec

### PR DESCRIPTION
**Pull request by sershe**

### CODE REVIEW

File: java/core/src/java/org/apache/orc/impl/WriterImpl.java [hunk 1]
```diff
@@ -645,6 +645,8 @@ public void appendUserMetadata(List<OrcProto.UserMetadataItem> userMetadata) {
     return ReaderImpl.deserializeStats(builder.getStatisticsList());
   }
 
+  // TODO: remove this
```
Comment (omalley) [inline java/core/src/java/org/apache/orc/impl/WriterImpl.java]: Remove the changes in this file. We do not want a guice dependency.

File: java/core/src/java/org/apache/orc/impl/ReaderImpl.java [hunk 2]
```diff
@@ -472,8 +473,9 @@ public static OrcTail extractFileTail(ByteBuffer buffer, long fileLength, long m
         .setPostscript(ps)
         .setFooter(footer)
         .setFileLength(fileLength);
+      isCodecError = false;
     } finally {
```
Comment (omalley) [inline java/core/src/java/org/apache/orc/impl/ReaderImpl.java]: try {} finally is pretty broken. You are better off using the try-with-resources statement.

File: java/core/src/java/org/apache/orc/impl/RecordReaderImpl.java [hunk 3]
```diff
@@ -1353,6 +1353,8 @@ public static String encodeTranslatedSargColumn(int rootColumn, Integer indexInS
     return result;
   }
 
+  // TODO: remove this
```
Comment (omalley) [inline java/core/src/java/org/apache/orc/impl/RecordReaderImpl.java]: Revert the changes in this file.

**Comment by sershe** [comment]
Updated the JIRA with motivation. 
I will rename the flag; there isn't a better way primarily because of the TODOs you are asking me to remove - we don't have access to information about errors if the user and the return-er of the codec are two different pieces of code at different times.
There's also no good way to check the state of the codec to see if it's valid or not. Hadoop ZLib codec may just throw NPE in reset when it's invalid, so the "Safely" catch block will handle that case. However it's nice to give user an opportunity to be proactive about this, esp. in the case of the get-use-return usage pattern

File: java/core/src/java/org/apache/orc/impl/ReaderImpl.java (hunk 2 above)
Comment (sershe) [reply java/core/src/java/org/apache/orc/impl/ReaderImpl.java]: existing code

File: java/core/src/java/org/apache/orc/impl/WriterImpl.java (hunk 1 above)
Comment (sershe) [reply java/core/src/java/org/apache/orc/impl/WriterImpl.java]: Removed the attribute, clarified the TODO

File: java/core/src/java/org/apache/orc/impl/RecordReaderUtils.java [hunk 4]
```diff
@@ -306,15 +330,24 @@ public void releaseBuffer(ByteBuffer buffer) {
...
     public DataReader clone() {
+      if (this.file != null) {
+        throw new UnsupportedOperationException(
+            "Cannot clone a DataReader that is already opened");
+      }
       try {
-        return (DataReader) super.clone();
+        DefaultDataReader clone = (DefaultDataReader) super.clone();
+        // Make sure we don't share the same codec between two readers.
+        clone.codec = OrcCodecPool.getCodec(clone.compressionKind);
```
Comment (prasanthj) [inline java/core/src/java/org/apache/orc/impl/RecordReaderUtils.java]: can this be moved inside super.clone()?

Comment (sershe) [reply java/core/src/java/org/apache/orc/impl/RecordReaderUtils.java]: super.clone() is Object.clone(), so no
